### PR TITLE
I DECLARE BANKRUPTCY

### PR DESCRIPTION
**Pull request by northeastprince**
@sampoder

### CODE REVIEW

File: app/models/hackathon/digest/listings.rb [hunk 1]
```diff
@@ -7,24 +7,32 @@ module Hackathon::Digest::Listings
...
       .flat_map { |criterion| criterion.new(recipient:).candidate_listings }
       .sort_by { |candidate| candidate[:hackathon].starts_at }
       .first(max_listings)
   end
 
-  def build_candidate_listings
-    applicable_listings.each do |result|
-      listings.build hackathon: result[:hackathon], subscription: result[:subscription]
+  def minimum_one_unseen_listing
+    if listed_hackathons.any? ->(hackathon) { Hackathon::Digest::Listing.exists?(recipient:, hackathon:) }
```
Comment (northeastprince) [inline app/models/hackathon/digest/listings.rb:34]:
```suggestion
    if listed_hackathons.all? ->(hackathon) { Hackathon::Digest::Listing.exists?(recipient:, hackathon:) }
```

File: app/models/hackathon/digest/listings.rb [hunk 2]
```diff
@@ -7,24 +7,32 @@ module Hackathon::Digest::Listings
...
   private
 
   LISTING_CRITERIA = [Criterion::Location]
   MAX_LISTINGS = 8
 
+  def build_candidate_listings
+    applicable_listings.each do |result|
+      listings.build hackathon: result[:hackathon], subscription: result[:subscription]
+    end
+  end
```
Comment (sampoder) [inline app/models/hackathon/digest/listings.rb:24]: @northeastprince I'm thinking about how we can do the filtering in the mailer, what would be the best way to add a field to a listing that is like "new = true / false"? that'd make filtering super easy in the view and we can use the logic in `minimum_one_unseen_listing` to conditionally apply set the tag?

Comment (northeastprince) [reply app/models/hackathon/digest/listings.rb:24]: How's this?

```ruby
# Hackathon::Digest::Listing

def previously_listed_for_recipient?(recipient: digest.recipient) do
  exists?(hackathon: hackathon, recipient: recipient)
end
```

Comment (sampoder) [reply app/models/hackathon/digest/listings.rb:24]: I was more thinking of something like a database field, would that make sense here? I want for when we build listings us to be able to differentiate between new ones and old / stale ones

Comment (sampoder) [reply app/models/hackathon/digest/listings.rb:24]: alright, pushed!

File: app/models/hackathon/digest/listing.rb [hunk 3]
```diff
@@ -2,4 +2,8 @@ class Hackathon::Digest::Listing < ApplicationRecord
   belongs_to :digest, class_name: "Hackathon::Digest"
   belongs_to :hackathon
   belongs_to :subscription
+
+  def previously_listed_for_recipient?(recipient: digest.recipient)
+    Hackathon::Digest::Listing.where(hackathon: hackathon, recipient: recipient) > 1
```
Comment (northeastprince) [inline app/models/hackathon/digest/listing.rb:7]:
```suggestion
    self.class.where(hackathon:, recipient:) > 1
```

File: app/views/hackathons/digest_mailer/digest.html.erb [hunk 4]
```diff
@@ -1,13 +1,29 @@
...
+  <% if (count = @digest.listings.count { |listing| listing.previously_listed_for_recipient? }) > 1 %>
+    these <%= count %> newly listed hackathons
   <% else %>
-    this hackathon
+    this newly listed hackathon
   <% end %>
   coming up near you.
 </p>
 
 <% @listings_by_subscription.each do |subscription, listings| %>
```
Comment (northeastprince) [inline app/views/hackathons/digest_mailer/digest.html.erb:11]: I would say maybe set these in the mailer so you don't have to repeat the `reject`s? Like `@new_listings_by_subscription` and `@previous_listings_by_subscription`

File: app/views/hackathons/digest_mailer/digest.html.erb [hunk 5]
```diff
@@ -1,13 +1,29 @@
 <p>
   Check out
-  <% if (count = @digest.listings.count) > 1 %>
-    these <%= count %> hackathons
+  <% if (count = @digest.listings.count { |listing| listing.previously_listed_for_recipient? }) > 1 %>
+    these <%= count %> newly listed hackathons
```
Comment (northeastprince) [inline app/views/hackathons/digest_mailer/digest.html.erb:4]:
```suggestion
    these <%= count %> new hackathons
```

File: app/views/hackathons/digest_mailer/digest.html.erb [hunk 6]
```diff
@@ -1,13 +1,29 @@
...
   <% end %>
   coming up near you.
 </p>
 
-<% @listings_by_subscription.each do |subscription, listings| %>
+<% @new_listings_by_subscription.each do |subscription, listings| %>
   <%= render "subscription", subscription:, listings: %>
 <% end %>
+
+<% if @previous_listings_by_subscription.count > 0 %>
```
Comment (northeastprince) [inline app/views/hackathons/digest_mailer/digest.html.erb:15]:
```suggestion
<% if @previous_listings_by_subscription.any? %>
```

File: app/mailers/hackathons/digest_mailer.rb [hunk 7]
```diff
@@ -7,6 +7,16 @@ def digest
       .includes(:subscription, hackathon: {logo_attachment: :blob})
       .group_by(&:subscription)
 
+    @new_listings_by_subscription = @digest.listings
+      .includes(:subscription, hackathon: {logo_attachment: :blob})
+      .select { |listing| listing.previously_listed_for_recipient? })
```
Comment (sampoder) [inline app/mailers/hackathons/digest_mailer.rb:12]:
```suggestion
      .select { |listing| listing.previously_listed_for_recipient? }
```

File: app/mailers/hackathons/digest_mailer.rb [hunk 8]
```diff
@@ -7,6 +7,16 @@ def digest
...
+    @new_listings_by_subscription = @digest.listings
+      .includes(:subscription, hackathon: {logo_attachment: :blob})
+      .select { |listing| listing.previously_listed_for_recipient? }
+      .group_by(&:subscription)
+
+    @previous_listings_by_subscription = @digest.listings
+      .includes(:subscription, hackathon: {logo_attachment: :blob})
+      .reject { |listing| listing.previously_listed_for_recipient? }
+      .group_by(&:subscription)
+
```
Comment (garyhtou) [inline app/mailers/hackathons/digest_mailer.rb:19]: Check out [`partition`](https://apidock.com/ruby/v2_5_5/Enumerable/partition) :)

**Comment by northeastprince** [comment]
Don't think this is needed with #352

**Pull request closed without merging**